### PR TITLE
[No QA] [TS migration] Migrate 'useKeyboardShortcut.js' hook to TypeScript

### PR DESCRIPTION
**Pull request by kubabutkiewicz**
### Fixed Issues

$ https://github.com/Expensify/App/issues/24942

### Tests

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Login into app
2. Use shortcut CMD + J
3. Use shortcut CMD + K
4. Use shortcut CMD + SHIFT + K 
5. Check if every keyboard shortcut is working

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/25230417/1677127a-db46-44da-b119-04cd83a22df7

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/25230417/aafc000e-82f3-4ae3-b8f8-9de5e9981e6a

</details>

### CODE REVIEW

File: src/hooks/useKeyboardShortcut.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import {useEffect} from 'react';
+import {ValueOf} from 'type-fest';
+import KeyboardShortcut from '@libs/KeyboardShortcut';
+import CONST from '@src/CONST';
+
+type Shortcut = ValueOf<typeof CONST.KEYBOARD_SHORTCUTS>;
+type KeyboardShortcutConfig = {
+    captureOnInputs?: boolean;
```
Comment (VickyStash) [inline src/hooks/useKeyboardShortcut.ts:8]: @kubabutkiewicz Maybe you can add comments, you can reuse [this](https://github.com/Expensify/App/blob/d7b14792974b4082ab574e7772dc8f036ffe38b0/src/libs/KeyboardShortcut/index.ts#L124) I guess

File: src/libs/KeyboardShortcut/index.ts [hunk 2]
```diff
@@ -130,12 +132,12 @@ function subscribe(
     key: string,
     callback: () => void,
     descriptionKey: string,
-    modifiers: string[] = ['shift'],
+    modifiers: ShortcutModifiers = ['CTRL'],
```
Comment (VickyStash) [inline src/libs/KeyboardShortcut/index.ts:135]: @kubabutkiewicz Why did you change the default value?

Comment (kubabutkiewicz) [reply src/libs/KeyboardShortcut/index.ts:135]: @VickyStash Because modifiers are declared [here](https://github.com/Expensify/App/blob/a31933a049b9a9aa4222a4d3edab19b54a4555b7/src/CONST.ts#L302C1-L302C1) and there is no shortcut with only shift key modifier so it doesn't make sense to default it to that key

File: src/libs/KeyboardShortcut/index.ts [hunk 3]
```diff
@@ -130,12 +132,12 @@ function subscribe(
...
     captureOnInputs = false,
     shouldBubble = false,
     priority = 0,
     shouldPreventDefault = true,
-    excludedNodes = [],
+    excludedNodes: string[] | readonly never[] = [],
```
Comment (VickyStash) [inline src/libs/KeyboardShortcut/index.ts:140]: @kubabutkiewicz Could you please clarify why `readonly never[]` is necessary?

Comment (kubabutkiewicz) [reply src/libs/KeyboardShortcut/index.ts:140]: @VickyStash It because of that
image
we are setting default value which comes from `CONST` and its basically an empty array but kt comes from `CONST` thats why its need to be typed as `readonly never[]`

File: src/hooks/useKeyboardShortcut.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            shortcut.descriptionKey ?? '',
+            shortcut.modifiers,
+            captureOnInputs,
+            shouldBubble,
+            priority,
+            shouldPreventDefault,
+            excludedNodes,
+        );
+
+        return () => {
```
Comment (barttom) [inline src/hooks/useKeyboardShortcut.ts:57]: or `return unsubscribe;`
also i think taht should work - little bit shorter version

```
return KeyboardShortcut.subscribe(
            shortcut.shortcutKey,
            callback,
            shortcut.descriptionKey ?? '',
            shortcut.modifiers,
            captureOnInputs,
            shouldBubble,
            priority,
            shouldPreventDefault,
            excludedNodes,
        );
```

**Review by VickyStash — APPROVED** [review]

File: src/hooks/useKeyboardShortcut.ts (hunk 4 above)
Comment (kubabutkiewicz) [reply src/hooks/useKeyboardShortcut.ts:57]: I would prefer no changing it to be sure to not introduce regressions, theoretically there shouldn't be be we never know 😅

File: src/libs/KeyboardShortcut/index.ts (hunk 3 above)
Comment (blazejkustra) [inline src/libs/KeyboardShortcut/index.ts:140]:
```suggestion
    excludedNodes: string[] = [],
```

File: src/libs/KeyboardShortcut/index.ts [hunk 5]
```diff
@@ -151,7 +153,7 @@ function subscribe(
         captureOnInputs,
         shouldPreventDefault,
         shouldBubble,
-        excludedNodes,
+        excludedNodes: [...excludedNodes],
```
Comment (blazejkustra) [inline src/libs/KeyboardShortcut/index.ts:156]:
```suggestion
        excludedNodes,
```

File: src/libs/KeyboardShortcut/index.ts [hunk 6]
```diff
@@ -102,13 +104,13 @@ function unsubscribe(displayName: string, callbackID: string) {
...
  */
-function getPlatformEquivalentForKeys(keys: string[]): string[] {
+function getPlatformEquivalentForKeys(keys: ShortcutModifiers): string[] {
     return keys.map((key) => {
         if (!(key in CONST.PLATFORM_SPECIFIC_KEYS)) {
             return key;
         }
 
-        const platformModifiers = CONST.PLATFORM_SPECIFIC_KEYS[key as keyof typeof CONST.PLATFORM_SPECIFIC_KEYS];
+        const platformModifiers = CONST.PLATFORM_SPECIFIC_KEYS[key];
```
Comment (blazejkustra) [inline src/libs/KeyboardShortcut/index.ts:113]: Nice one 👍

File: src/hooks/useKeyboardShortcut.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        const unsubscribe = KeyboardShortcut.subscribe(
+            shortcut.shortcutKey,
+            callback,
+            shortcut.descriptionKey ?? '',
+            shortcut.modifiers,
+            captureOnInputs,
+            shouldBubble,
+            priority,
+            shouldPreventDefault,
+            excludedNodes,
```
Comment (blazejkustra) [inline src/hooks/useKeyboardShortcut.ts:54]:
```suggestion
            excludedNodes as string[],
```

File: src/hooks/useKeyboardShortcut.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    /* Do not capture key events targeting excluded nodes (i.e. do not prevent default and let the event bubble) */
+    excludedNodes?: string[];
+    /* Is keyboard shortcut is already active */
+    isActive?: boolean;
+};
+
+/**
+ * Register a keyboard shortcut handler.
+ * Recommendation: To ensure stability, wrap the `callback` function with the useCallback hook before using it with this hook.
+ *
```
Comment (blazejkustra) [inline src/hooks/useKeyboardShortcut.ts:25]:
```suggestion
```

**Review by blazejkustra — APPROVED** [review]
Only small remarks! LGTM

**Review by chiragsalian — APPROVED** [review]
Code LGTM, but would you be able to investigate the failed perf-tests GH action.

**Comment by kubabutkiewicz** [comment]
@chiragsalian perf test now passing
